### ISSUE TITLE
Load-balancing when there are multiple matching algorithm implementations

### ISSUE THREAD

**Issue by docularxu** [openssl/openssl#19228]
In a library context, there is a possibility that multiple implementations of an algorithm can exist. For example, a _libctx_ can load one provider which uses the underlying hardware accelerator, the same _libctx_ can then load another provider which links to a software accelerator library such as multi-buffer solutions based on vector instructions, and both providers can support Digest algorithms (MD5, etc).

From the system’s viewpoint, it is more efficient when user applications don't stick to one single implementation, even when property query strings are the same. For example, when hardware accelerator reaches its limit of bandwidth, it would be wise for OpenSSL/libcrypto to fetch and return the algorithm implementation from the software accelerator, if available.

Take a real world example from Ceph. [Ceph uses MD5 from OpenSSL for its object storage](https://github.com/ceph/ceph/blob/main/src/common/ceph_crypto.h). Ceph can issue hundreds’ of MD5 tasks simultaneously from its threads pool. You can bind Ceph to a hardware accelerator. But hardware has a hard limit in bandwidth. When this limit is passed, performance downgrades a lot. In such a case, it will be wise to “off-load” part of these MD5 tasks to other providers, such as the software accelerator.

In current OpenSSL 3.0 crypto, automatic ‘load-balancing’ among algorithm implementations is not possible. The user has to specify different property query strings in order to fetch different algorithm implementations. But why should the user care about the property string? When all the user cares about, from its original purpose, is the algorithm name. The user already delegates the algorithm implementation (and the selection of algorithm implementations) to OpenSSL crypto library.

So here I want to propose a solution to support this feature, which I call it ‘load balancing among providers’. 

PS, I know my idea is immature, and my understanding of OpenSSL design philosophy may be biased. But still, hopefully if I can get some feedback from the community and together we bring this discussion to a deeper level. Appreciate your time and effort.

**Here comes my proposal**

Currently, in OpenSSL 3.0, there are two steps in binding a method (i.e. algorithm implementation) to an operation context. 1) type = APINAME_fetch(), which returns one qualified method. 2) APINAME_Init(ctx, type, …), which binds the method to the operation context. After these two steps, the user just use the ‘operation context’ to do its jobs and free it when jobs are done.

To support runtime load-balancing in library context, we can modify these two steps.

1. Algorithm Fetching. Define a new __reserved__ property name (such as ‘`load_balance=1`’) to highlight that a provider wants to join run-time load balancing. Then, when a property query string includes this property in _fetch, `APINAME_fetch(x, x, “load_balance=1”)` will traverse the method store and return a object containing a method list of all matching algorithm implementations. There is no need to change APINAME_fetch() prototype, the list can be embedded into the returned object struct, together with one field or two about load_balancing state.

1. At the operation initialization phase, APINAME_Init(ctx, type, …) (such as _EVP_DigestInit_ex_), one implementation will be selected from the list previously returned by algorithm fetching, and this implementation will be bound to the operation context (such as a digest context _EVP_MD_CTX *_). At the same time, load_balancing state will be updated for next call.

There are other design points which need to be considered.

1. Strategies for selecting one method from all qualified implementations need to be defined. This can be configured at library context level, eg. loading through configuration file. Some common strategies can be pre-defined in OpenSSL, for example ‘round-robin’, ‘packet-size’, ‘failover’. In the long run, we can allow the provider author to define its own load-balancing hints, and use such hints to define customized strategy.

1. Caching of the algorithm fetching list. This can be saved to cache for future use, and improve performance further.

**Comment by mattcaswell** [comment]
An interesting idea.

> Algorithm Fetching. Define a new reserved property name (such as ‘load_balance=1’) to highlight that a provider wants to join run-time load balancing

A provider may not know that it should be considered for load balancing. This is more likely to be a characteristic determined at deployment time, rather than something determined by the provider author. I would have thought that the list of providers eligible for load balancing should be specified in the config file, rather than by the provider author.

**Comment by paulidale** [comment]
Agreed, this is an interesting concept.

The non-deterministic result of the fetch functions is deliberate as is the early binding upon fetch (in reality, it is deterministic, just undefined).  Late binding is supported via the myriad `EVP_aes_128_gcm()` style calls but they impose a significant performance penalty when used.  We've also skirted the issue of algorithm priority -- we're almost certainly going to need it in due course but not just yet (it's difficult).

**Comment by mattcaswell** [comment]
I think it might actually be feasible to implement something now using released 3.0, i.e. via a 3rd party "load balancing provider". Such a provider would have to have its own internal libctx into which it loads the various providers it is is fronting. It would also have to "implement" the algorithms it wants to balance - but that implementation would just be a thin proxy implementation. When a new operation using the algorithm is "inited" it would select a provider implementation from its available providers and "forward" all calls to the algorithm implementation from that point onwards. The end application would simply have to ensure it fetched its alg implementation from the load balancing provider and would have no visibility of which underlying provider is actually supplying the implementation.

**Comment by docularxu** [comment]
Based on @mattcaswell 's comment 
> Such a provider would have to have its own internal libctx into which it loads the various providers it is is fronting.

I made this proposal. It provides a draft about which structures and which functions need to be touched and why. Sorry for the long texts, but many thanks for your reading. And your feedback is appreciated.

In general, there will be a ‘load balancing’ provider. This provider creates its own child libctx, and the child libctx can load / unload its own providers, and get mirror‘ed providers from the application libctx. Then, for ‘load balancing’, we add:

### “Load balancing” provider:
```
OSSL_provider_init() {
 - Create a child libctx
 - Register scheduler_cb() into child libctx’s method store.
              with Prototype: scheduler_cb(alg->impls, scheduler_status, &method)
 - Create and initialize a scheduler_status, who contains all status information a scheduler
             need to know to make decisions, and
 - Register this scheduler_status into child libctx’s method store.
}
```

My thought is one provider can provide only one load balancing strategy, although the detailed parameters (such as thresholds) may change. When people designed a new strategy, they need to contribute a new provider which has new scheculer_cb() and new scheduler_status.

### Method Store / struct ossl_method_store_st:
To facilitate the load balancing scheduling, add two new fields into the method_store structure. Both fields comes from the “load balancing” provider when available, or NULL.

```
 - a scheduler_cb() function, who enforces the scheduling strategy. 
 - a scheduler_status field, usually a pointer, who points to a structure that contains 
   all status information a scheduler need to know to make scheduling decisions.
```

### At algorithm fetching phase,
In _fetch(application_libctx, char * algorithm, “load-balancing”):
 - Return a method created from “load balancing” provider’s OSSL_ALGORITHM [ ].

### At the operation initialization phase,
In APINAME_Init (ctx, type, ...):
, When (`type->prov == “load-balancing”`), then call `_fetch(child_libctx, ….) `to search through child providers in Child Library Context, 

 - select an implementation, fetch the method object from cache or from the method store, then use this method to initialize the operation context:
 -- Create a new algorithm implementation context, ie. algctx
 -- Call this method’s initialization function. eg. digest->dInit(algctx, params)

There are two paths for _fetch’ing a method from any libctx: one is through cache, the other is through method store, i.e.  ossl_method_construct(). For load-balancing strategy's purpose, we need to insert the scheduling strategy into both paths.

### Path 1: Cache
#### struct QUERY:
Add this logic into QUERY. When (`QUERY.method->method == NULL`), that means there is no method selected for the given *query and *provider. It further means this is a load-balancing operation required.

#### ossl_method_store_cache_get(store, prov, meth_id, propq, &method):
When the cache item exists, and (method->method == NULL), call 
`store->scheduler_cb(alg->impls, store->scheduler_status, &method)
`
#### ossl_method_store_cache_set(...):
When this is a result of ‘load-balancing’, set `method->method `to `NULL`.

### Path 2: ossl_method_construct()
Down this path, there is no change in ‘method store’ construction, no matter it’s a temporary store or a permanent store. The load-balancing logic should be inserted after the ‘method store’ is constructed, ie., into ossl_method_store_fetch().

#### ossl_method_store_fetch(*store, meth_id, methdata->prop_query, prov, &method):
Check whether the `prop_query` contains “load-balancing”, if yes call 
`store->scheduler_cb(alg->impls, store->scheduler_status, &method)
`

There is one more thing TODO: the store can be either a temporary store, or it can be the child libctx’s permanent store. Need to add scheduler_cb() and scheduler_status into the temporary store when it’s created. I think they can be copied from child libctx when we need them.

Reference: `alg = ossl_method_store_retrieve(store, nid)`

**Comment by docularxu** [comment]
I just submitted a request for comments version, https://github.com/openssl/openssl/pull/19706 . Design details in the pull request. Your comments of any type will be highly appreciated.

The design is based on child libctx. On top of which, I defined a new type of provider and libctx (namely load_balancer libctx). Per what @mattcaswell proposed, a wrapper (dummy) implementation is provided by "loadbalance" provider, and at Init stage (like DigestInit), the real implementation will be fetched from the load-balance scheduler.

`md5 = EVP_MD_fetch(parent_libctx, "MD5", "provider=loadbalance");
`
`EVP_DigestInit_ex(ctx, md5, NULL);
`

Then at the init stage, i.e. digest init, the evp code will detect that this method comes from loadbalancer provider, so, it will go down one level, ie. check through the load_balancer libctx’s method store or cache, call previously bounded scheduler function to fetch a proper implementation from child providers, then set the new fetched method into context ctx.
